Add recipe for calctape

Boruch-Baum opened the pull request:
### Brief summary of what the package does

See the animated gif screencast on the repo home page.

This package provides an easy-to-use tape calculator functionality natively and intuitively IN ANY BUFFER. You can use it to sum an existing column of numbers, without even having to select the entire region, and even if the numbers are terribly aligned. You can use it to create a “tape” in ANY writable buffer, and can include annotations for each line. It auto-aligns its tapes, supports memory operations, can perform VAT / sales tax calculations, and seamlessly handles integers, floats, and scientific notation on the same tape. It also sneaks in a wrapper around Emacs quick-calc that formats output with thousands delimiters.

### Direct link to the package repository

https://github.com/Boruch-Baum/emacs-calctape

### Your association with the package

Author, developer, maintainer.

### Relevant communications with the upstream package maintainer

N/A

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

NOTES:
1) Four byte-compile warnings were encountered. The first was a placeholder definition used to keep the code organized. The other three were false claims of unused lexical variables
2) Checkdoc gave some unusual recommendations, including suggestions to change variable names, flagging words such as 'docstring', 'whitespace'.

riscy commented:
Thanks for this.  I see your NOTES above (four byte-compiler warnings, checkdoc).  The byte-compiler warnings are almost always something we want to fully address.  Checkdoc is "within reason" - not strict and I wouldn't bother for your package since it looks fine.  Here's a quick first pass -

⸺ `calctape.el` with byte-compile using Emacs 30.1:
```
calctape.el:382:8: Warning: function `calctape--validate-symbols' defined multiple times in this file
In calctape--make-box:
calctape.el:691:9: Warning: Unused lexical variable `start-point-marker'
calctape.el:743:11: Warning: assignment to free variable `line-end-column'
calctape.el:746:10: Warning: reference to free variable `line-end-column'
In calctape--find-number-string:
calctape.el:860:9: Warning: Unused lexical variable `current-num-begin-pos'
In calctape--edit:
calctape.el:1317:13: Warning: assignment to free variable `return-list'
calctape.el:1325:36: Warning: reference to free variable `return-list'
In calctape--create:
calctape.el:1437:8: Warning: Unused lexical variable `thou-len'
```

⸺ `calctape.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- You can separate TODOs from package `;;; Commentary:` with a `;;; TODO:` section
- calctape.el#L155: Only use back/front quotes to link to top-level elisp symbols
- calctape.el#L977: Only use back/front quotes to link to top-level elisp symbols
- calctape.el#L778: Consider `beginning-of-line`
- calctape.el#L800: Consider `beginning-of-line`
- calctape.el#L807: Consider `beginning-of-line`
- calctape.el#L867: Consider `beginning-of-line`
- calctape.el#L933: Consider `beginning-of-line`
- calctape.el#L569: Consider `end-of-line`
- calctape.el#L743: Consider `end-of-line`
- calctape.el#L747: Consider `end-of-line`
- calctape.el#L760: Consider `end-of-line`
- calctape.el#L772: Consider `end-of-line`
- calctape.el#L775: Consider `end-of-line`
- calctape.el#L794: Consider `end-of-line`
- calctape.el#L797: Consider `end-of-line`
- calctape.el#L1243: Consider `end-of-line`
- calctape.el#L1768: Consider `end-of-line`
- calctape.el#L1783: Consider `end-of-line`
```

Boruch-Baum commented:
Thanks for the second look. I just now pushed commit 84242d4 which
addresses most of the issues (see below for proofs of the false
positives). The melpazoid catches were nice, thanks. Some of what the
byte-compiler reported were real and I corrected them. However,

> calctape.el:382:8: Warning: function `calctape--validate-symbols'
> defined multiple times in this file
That was intentional to be able to group the functions in a sane
way for readability.


Here are some outright false positives:

> calctape.el:691:9: Warning: Unused lexical variable `start-point-marker'
But see that it *is* used on line 699

> calctape.el:1437:8: Warning: Unused lexical variable `thou-len'
But see that it *is* used on line 1591

> calctape.el:860:9: Warning: Unused lexical variable `current-num-begin-pos'
But see that it *is* used on line 877

-- 
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

Boruch-Baum commented:
I've also posted a second commit, to integrate melpazoid as a github
workflow, per the instructions on your repo page. The results aren't
instantaneous ... it has been "in progress" for several minutes now.

-- 
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

Boruch-Baum commented:
I appreciate the extra time you're spending on this. I've pushed an
updated commit that removes the two "Unused lexical variable" (`let',
`setq'-only) errors, uses `declare-function' instead of a empty `defun'
stub, and uses \=' notation in docstrings for single quotes that
surround things other than elisp functions.

I double-checked the consequences of using `defun' or `declare-function'
in the file, positioned as it is, above the `defcustom' that invokes it,
and it doesn't exhibit the problems you anticipate: 1) The `defcustom'
does correctly invalidate bad assignment attempts, so it is using the
correct function, not the stub (ie. even when using `defun' instead of
`declare-function'); 2) M-x `describe-function' does use the correct
docstring.


On 2025-09-28 09:25, Chris Rayner wrote:
>
> ...
>
>    Re: declaring a function twice - don't do this. If you need to
>    forward-declare a function, I think you should be using
>    declare-function. But you should know that the defcustom is not just
>    referring to the function here, it is actually calling it, so it's
>    calling and using the value returned by your stub:
>
> ...
>
>    I think this'll also create confusion with the online help (if a user
>    clicks the help for calctape--validate-symbols and clicks to view the
>    source, it looks like they'll be sent to your stub definition).
>
>    You also mention that the "unused lexical variable" warnings are not
>    correct. I'll just point out one of these:
>
> ...
>
>    The reason start-point-marker is being complained about is because you
>    bind start-point-marker, and then you setq start-point-marker, but you
>    never make use of the value that you're setting it to.

-- 
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

riscy commented:
>  I'd like to install melpazoid locally, but it would be my first use of docker, and I'm resistant to accepting the overhead on this old/limited desktop.

The docker requirement is actually CRAZY - I realize this. :)  It was an "easy" tool for me to reach for when I started looking for a way to test a package on someone else's computer, but it really is a big requirement for people to run locally.

> Using `declare-function' causes `load-file' to fail

wrt to this, I notice that you added a comment in your code that says 
```
although `eval-buffer' does work.
```
Note this isn't true about `eval-buffer` (I just confirmed) - I imagine what's happening is that your code is already loaded into the runtime of the Emacs instance you were testing with, and so evaluating the buffer is just using existing definitions in the obarray during evaluation.

Altogether I think this (and my original comments above) are still legitimate:

⸺ `calctape.el` with byte-compile using Emacs 30.1:
```
calctape.el:387:8: Warning: function `calctape--validate-symbols' defined multiple times in this file
```

Boruch-Baum commented:
On 2025-10-05 09:56, Chris Rayner wrote:
>    wrt to this, I notice that you added a comment in your code that says
> although `eval-buffer' does work.
>
>    Note this isn't true about eval-buffer (I just confirmed) - I imagine
>    what's happening is that your code is already loaded into the runtime
>    of the Emacs instance you were testing with, and so evaluating the
>    buffer is just using existing definitions in the obarray during
>    evaluation.

I double-checked now and you are correct about the eval-buffer.

However, the more-pertinent point does remain that `load-file' does fail
when using `declare-function' but `load-file' does succeed when using
`defun' with the dummy contents. And, that using `defun' with the dummy
contents does not adversely impact the subsequent `defcustom'.

The only reason for being stubborn about this is readability and
organization of the code. I want to keep the function definitions in a
single location, separate from the variable definitions.

If that's a deal-breaker, fine, I can concede the point, but in the
bigger picture, as a policy, IMO it detracts from code organization and
readability.

-- 
hkp://keys.gnupg.net
CA45 09B5 5351 7C11 A9D1  7286 0036 9E45 1595 8BC0

Boruch-Baum commented:
Done.

Also, I wrote a script to help me run melpazoid locally. Then I generalized it a bit. The following version needs the user to manually define MELPAZOID_PATH to the directory in which melpazoid's make file is located.

```
#!/bin/bash
# melpazoid-local.sh
case "${1}" in
--help|-h|-?|--usage)
   printf "\n$0 [path-to-repo-root [recipe-filename]]\n
Helper script for running melpazoid locally.\n
What that means: You have docker installed, probably running 'rootless',
and you want to perform the melpazoid lint-checking routine on a local
elisp package repository for which you have a melpa 'recipe' written.
If you haven't previously downloaded the melpazoid docker image, this
process will need to do that (about 700 MB disk space).\n\n"
   exit
   ;;
esac

MELPAZOID_PATH=

LOCAL_REPO="${1}"
RECIPE_PATH="${2}"

SERVICE_MANAGER=$(ps -p 1 -o comm=)
case ${SERVICE_MANAGER} in
systemd)
  if ! systemctl is-active --quiet docker.service; then
    echo "Docker service is not running."
    exit
  fi
  ;;
init)
  if ! service docker status > /dev/null 2>&1; then
    echo "Docker service is not running."
    exit
  fi
  ;;
esac

LOCAL_REPO="${LOCAL_REPO/#\~/$HOME}"
while [ ! -d "${LOCAL_REPO}" ] ; do
  read -ep "Enter path to local repository root: " LOCAL_REPO
  LOCAL_REPO="${LOCAL_REPO/#\~/$HOME}"
done
if [[ "${LOCAL_REPO: -1}" != "/" ]]; then
  LOCAL_REPO="${LOCAL_REPO}/"
fi
export LOCAL_REPO="${LOCAL_REPO}"

while [ ! -f "${LOCAL_REPO}${RECIPE_PATH}" ] ; do
  read -ep "Enter path to package recipe file (relative to root): " RECIPE_PATH
  done
export RECIPE=$( < "${LOCAL_REPO}${RECIPE_PATH}" )

if ! docker images |grep "^melpazoid" &>/dev/null ; then
  printf "ATTENTION: No local melpazoid image found !\n\n
That means the melpazoid program will have docker build one.
Expect this to take some time, but the image should persist
locally, so subsequent runs will execute much quicker.\n\n"
fi

OUT_FILE="${LOCAL_REPO}melpazoid_output_$( printf "%(%F_%H%M%S)T" ).txt"
printf "Running.\nOutput is being re-directed to:\n${OUT_FILE}\n..."
echo "${OUT_FILE}\n" > "${OUT_FILE}"
exec >>"${OUT_FILE}" 2>&1
cd "$MELPAZOID_PATH"
make
exec > /dev/tty 2>/dev/tty
printf "\ndone!\n\nPress Ctrl-C to exit,
or press any other key to view the output using 'less'.\n\n"
less "${OUT_FILE}"
```
